snapshot/storage/metastore.go

```diff
@@ -0,0 +1,69 @@
+package storage
+
+import (
+	"context"
+
+	"github.com/docker/containerd/snapshot"
+)
+
+// MetaStore is used to store metadata related to a snapshot driver. The
+// MetaStore is intended to store metadata related to name, state and
+// parentage. Using the MetaStore is not required to implement a snapshot
+// driver but can be used to handle the persistence and transactional
+// complexities of a driver implementation.
+type MetaStore interface {
+	// TransactionContext creates a new transaction context.
+	TransactionContext(ctx context.Context, writable bool) (context.Context, Transactor, error)
+
+	// Stat returns the snapshot stat Info directly from
+	// the metadata.
+	Stat(ctx context.Context, key string) (snapshot.Info, error)
+
+	// Walk iterates through all metadata for the stored
+	// snapshots and calls the provided function for each.
+	Walk(ctx context.Context, fn func(context.Context, snapshot.Info) error) error
+
+	// CreateActive creates a new active snapshot transaction referenced by
+	// the provided key. The new active snapshot will have the provided
+	// parent. If the readonly option is given, the active snapshot will be
+	// marked as readonly and can only be removed, and not committed. The
+	// provided context must contain a writable transaction.
+	CreateActive(ctx context.Context, key, parent string, readonly bool) (Active, error)
+
+	// GetActive returns the metadata for the active snapshot transaction
+	// referenced by the given key.
+	GetActive(ctx context.Context, key string) (Active, error)
+
+	// Remove removes a snapshot from the metastore. The provided context
+	// must contain a writable transaction. The string identifier for the
+	// snapshot is returned as well as the kind.
+	Remove(ctx context.Context, key string) (string, snapshot.Kind, error)
+
+	// Commit renames the active snapshot transaction referenced by `key`
```

Comment:
I think we need to update the [design doc](https://github.com/docker/containerd/blob/master/design/snapshots.md) again with this in mind, since that's still calling out a non-terminal `Commit`.  Not a blocker for merging, but we should at least have an issue that outlines what changes need to be made to the design.